File: app/src/main/java/com/buap/stu/buapstu/PrincipalActivity.java
package com.buap.stu.buapstu;

import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.buap.stu.buapstu.models.Alumno;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public class PrincipalActivity extends AppCompatActivity {
    Button LogOut, transferir;
    ImageButton addCreditos;
    TextView TextCreditos;
    private FirebaseAuth mAuth;
    private FirebaseUser User;
    DatabaseReference mDatabase;
    Alumno usuario;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {


        super.onCreate(savedInstanceState);

        mDatabase = FirebaseDatabase.getInstance().getReference();

        setContentView(R.layout.activity_principal);

        Bundle parametros = this.getIntent().getExtras();
        usuario = (Alumno) parametros.getSerializable("usuario");
        TextCreditos = (TextView) findViewById(R.id.textCreditos);
        mostrarCreditos();

        addCreditos = (ImageButton) findViewById(R.id.addCredit);

        LogOut = (Button) findViewById(R.id.CerrarSesion);
        transferir = (Button) findViewById(R.id.TC);
        mAuth = FirebaseAuth.getInstance();
        User = mAuth.getCurrentUser();

        LogOut.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                mAuth.signOut();
                Toast.makeText(PrincipalActivity.this, "Ha cerrado sesion", Toast.LENGTH_SHORT).show();
                Intent intent = new Intent(PrincipalActivity.this,LoginActivity.class);
                startActivity(intent);
            }
        });

        addCreditos.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                usuario.setCreditos(usuario.getCreditos()+10);
                mDatabase.child("Usuarios").child(usuario.getUid()).setValue(usuario);
                mostrarCreditos();
            }
        });

        transferir.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(PrincipalActivity.this,TransferirActivity.class);
                startActivity(intent);
            }
        });
    }

    private void mostrarCreditos() {
        TextCreditos.setText("Saldo: "+String.valueOf(usuario.getCreditos()+"    "));
    }
}

File: app/src/main/java/com/buap/stu/buapstu/PrincipalActivity.kt
package com.buap.stu.buapstu

import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.buap.stu.buapstu.core.states.AuthState
import com.buap.stu.buapstu.core.utils.launchActivity
import com.buap.stu.buapstu.core.utils.showToast
import com.buap.stu.buapstu.databinding.ActivityPrincipalBinding
import com.buap.stu.buapstu.models.Alumno
import com.buap.stu.buapstu.presentation.AuthViewModel
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class PrincipalActivity : AppCompatActivity() {

    private var _binding: ActivityPrincipalBinding? = null
    private val binding: ActivityPrincipalBinding get() = _binding!!

    private val authViewModel: AuthViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = ActivityPrincipalBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initButtons()
        initFlows()
    }

    private fun initFlows() = with(binding) {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.RESUMED) {
                lifecycleScope.launch {
                    authViewModel.userState.collect { state ->
                        (state as? AuthState.Authenticated)?.let { user ->
                            (user.user as? Alumno)?.let { user ->
                                textCreditsStudents.text = "${user.creditos}"
                            }
                        }
                    }
                }
            }
        }
    }


    private fun initButtons() = with(binding) {
        buttonLogOut.setOnClickListener {
            authViewModel.signOut()
            showToast("Ha cerrado sesion")
            launchActivity(
                activity = LoginActivity::class.java,
                clearStack = true,
                finishActivity = true
            )
        }

        buttonsAddCredits.setOnClickListener {
            authViewModel.addCredits()
        }

        buttonSaveTravel.setOnClickListener {
            launchActivity(
                activity = ReservaActivity::class.java,
                finishActivity = false,
                clearStack = false
            )
        }


    }

    override fun onDestroy() {
        super.onDestroy()
        _binding=null
    }
}